// lzy-React/lib/V3/myReactCore/Interface.ts
import type { FiberNode } from './GlobalFiber'

//全局变量声明
declare global {
    interface Window {
        $$routeContainerFiber?: any
        $$lzyReactGlobal?: Global
    }
}

//!------------Ref容器结构--------------
type RefObject = { current: any } | null

//!--------------单个UseEffectHook结构-------------------
//todo 通过判断tag 'useEffect','useLayoutEffect' 来判断何种钩子  改变执行时机
interface UseEffectHook {
    hookFlags: string,
    index: number,
    memorizedState: any,
    next: any
}

//!------------单个Effect(EffectUpdater)结构--------------------
interface Effect {
    tag: string,
    create: Function,
    destory: Function | null,
    deps: any[] | null,
    next: Effect | null,
}

//!-------------单个useStateHook结构-----------------
interface UseStateHook {
    hookFlags: string,
    index: number,
    memorizedState: any,
    updateStateQueue: { pending: any },//! hook.queue中保存了需要执行的update
    next: any
}

//!-------------单个StateUpdater结构--------------------
interface StateUpdater {
    action: Function | any,
    next: StateUpdater | null
}


//!-------------全局需要的变量结构---------------
interface Global {
    workInprogressFiberNode: FiberNode | null,
    workInProgressHook: { currentHook: any },
    EffectList: { firstEffect: any, lastEffect: any, length: number },
    LayoutEffectList: { firstEffect: any, lastEffect: any, length: number },
    destoryEffectsArr: Effect[],
    renderTag: 'update' | 'mount',
    isFirstRenderDone: boolean,
    islzySyntheticEventSystemOpen: boolean // 是否开启合成事件系统
}

//! -------------element结构----------------
type ElementNode = BaseElementNode | TextElementNode

type BaseElementNode = {
    $$typeof: symbol
    tag: string
    fiber: FiberNode
    children: (ElementNode | TextElementNode)[]
    elementMap?: Map<any, ElementNode | TextElementNode>
    ref?: any //TODO 预备ref类型
    props?: Record<any, any>
    key?: string
    text?: string
}

type TextElementNode = {
    $$typeof: symbol
    tag: 'text'
    text: string
    fiber: FiberNode
    children?: (ElementNode | TextElementNode)[]
    elementMap?: Map<any, ElementNode | TextElementNode>
    ref?: any
    props?: any
    key?: string
}

export type { FiberNode, StateUpdater, Effect, UseStateHook, UseEffectHook, Global, ElementNode, TextElementNode, RefObject }
