// lzy-Ruzy/out/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const QueryClient_1 = require("./query-core/QueryClient");
const StateCache_1 = require("./store-core/StateCache");
class Ruzy {
    constructor() {
        this.store = undefined;
        this.queryStore = undefined;
    }
    initStates(initState) {
        this.store = new StateCache_1.StateCache(initState);
        return this.store.getAllState();
    }
    initQueries(initQuery) {
        this.queryStore = new QueryClient_1.QueryClient(initQuery);
        return this.queryStore.getAllQueryData();
    }
    useState(...key) {
        var _a;
        if (!this.store)
            return console.error('store未初始化');
        return (_a = this.store) === null || _a === void 0 ? void 0 : _a.useBaseState(...key);
    }
    setState(param) {
        var _a;
        if (!this.store)
            return console.error('store未初始化');
        return (_a = this.store) === null || _a === void 0 ? void 0 : _a.setBaseState(param);
    }
    getState(key) {
        var _a;
        if (!this.store)
            return console.error('store未初始化');
        return (_a = this.store) === null || _a === void 0 ? void 0 : _a.getState(key);
    }
    getAllState() {
        var _a;
        if (!this.store)
            return console.error('store未初始化');
        return (_a = this.store) === null || _a === void 0 ? void 0 : _a.getAllState();
    }
    useQuery(keys, queryFn, options) {
        if (!this.queryStore)
            return console.error('queryStore未初始化');
        return this.queryStore.useBaseQuery(keys, queryFn, options);
    }
    getQueryData(key) {
        if (!this.queryStore)
            return console.error('queryStore未初始化');
        return this.queryStore.getQueryData({ queryKey: [key] });
    }
    getAllQueryData() {
        if (!this.queryStore)
            return console.error('queryStore未初始化');
        return this.queryStore.getAllQueryData();
    }
}
exports.default = Ruzy;
