// lzy-React/out/V3/myEvent/EventPriority.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getEventPriority = exports.setCurrentUpdatePriority = exports.getCurrentUpdatePriority = exports.IdleEventPriority = exports.DefaultEventPriority = exports.ContinuousEventPriority = exports.DiscreteEventPriority = void 0;
const FiberLane_1 = require("./FiberLane");
//todo 维护了一个全局当前的优先级,会根据任务执行实时变化
let currentUpdatePriority = FiberLane_1.NoLane;
exports.DiscreteEventPriority = FiberLane_1.SyncLane; // 离散事件优先级(切片事件) 最高级
exports.ContinuousEventPriority = FiberLane_1.InputContinuousLane; // 持续事件   次高级
exports.DefaultEventPriority = FiberLane_1.DefaultLane; // 默认事件优先级
exports.IdleEventPriority = FiberLane_1.IdleLane; // 闲置事件优先级
function getCurrentUpdatePriority() {
    return currentUpdatePriority;
}
exports.getCurrentUpdatePriority = getCurrentUpdatePriority;
function setCurrentUpdatePriority(newPriority) {
    currentUpdatePriority = newPriority;
}
exports.setCurrentUpdatePriority = setCurrentUpdatePriority;
// 将不同event进行分类
function getEventPriority(domEventName) {
    switch (domEventName) {
        //! 无return时会返回2  也就是DiscreteEventPriority 
        //! 绝大部分用户事件优先级为2 最高级
        case 'cancel':
        case 'click':
        case 'close':
        case 'contextmenu':
        case 'copy':
        case 'cut':
        case 'auxclick':
        case 'dblclick':
        case 'dragend':
        case 'dragstart':
        case 'drop':
        case 'focusin':
        case 'focusout':
        case 'input':
        case 'invalid':
        case 'keydown':
        case 'keypress':
        case 'keyup':
        case 'mousedown':
        case 'mouseup':
        case 'paste':
        case 'pause':
        case 'play':
        case 'pointercancel':
        case 'pointerdown':
        case 'pointerup':
        case 'ratechange':
        case 'reset':
        case 'resize':
        case 'seeked':
        case 'submit':
        case 'touchcancel':
        case 'touchend':
        case 'touchstart':
        case 'volumechange':
        // Used by polyfills:
        // eslint-disable-next-line no-fallthrough
        case 'change':
        case 'selectionchange':
        case 'textInput':
        case 'compositionstart':
        case 'compositionend':
        case 'compositionupdate':
        // Only enableCreateEventHandleAPI:
        // eslint-disable-next-line no-fallthrough
        case 'beforeblur':
        case 'afterblur':
        // Not used by React but could be by user code:
        // eslint-disable-next-line no-fallthrough
        case 'beforeinput':
        case 'blur':
        case 'fullscreenchange':
        case 'focus':
        case 'hashchange':
        case 'popstate':
        case 'select':
        case 'selectstart':
            return exports.DiscreteEventPriority;
        case 'drag':
        case 'dragenter':
        case 'dragexit':
        case 'dragleave':
        case 'dragover':
        case 'mousemove':
        case 'mouseout':
        case 'mouseover':
        case 'pointermove':
        case 'pointerout':
        case 'pointerover':
        case 'scroll':
        case 'toggle':
        case 'touchmove':
        case 'wheel':
        // Not used by React but could be by user code:
        // eslint-disable-next-line no-fallthrough
        case 'mouseenter':
        case 'mouseleave':
        case 'pointerenter':
        case 'pointerleave':
            return exports.ContinuousEventPriority;
        case 'message':
            return exports.DiscreteEventPriority;
        default:
            return exports.DefaultEventPriority;
    }
}
exports.getEventPriority = getEventPriority;
