// lzy-React/out/V3/myEvent/laneAndPriority.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IdleEventPriority = exports.DefaultEventPriority = exports.ContinuousEventPriority = exports.DiscreteEventPriority = exports.OffscreenLane = exports.IdleLane = exports.IdleHydrationLane = exports.SelectiveHydrationLane = exports.SomeRetryLane = exports.SyncUpdateLanes = exports.DefaultLane = exports.DefaultHydrationLane = exports.InputContinuousLane = exports.InputContinuousHydrationLane = exports.SyncLane = exports.SyncHydrationLane = exports.NoLane = exports.NoLanes = exports.TotalLanes = void 0;
// Lane values below should be kept in sync with getLabelForLane(), used by react-devtools-timeline.
// If those values are changed that package should be rebuilt and redeployed.
exports.TotalLanes = 31;
exports.NoLanes = 0b0000000000000000000000000000000;
exports.NoLane = 0b0000000000000000000000000000000;
exports.SyncHydrationLane = 0b0000000000000000000000000000001;
exports.SyncLane = 0b0000000000000000000000000000010;
exports.InputContinuousHydrationLane = 0b0000000000000000000000000000100;
exports.InputContinuousLane = 0b0000000000000000000000000001000;
exports.DefaultHydrationLane = 0b0000000000000000000000000010000;
exports.DefaultLane = 0b0000000000000000000000000100000;
exports.SyncUpdateLanes = 0b0000000000000000000000000101010;
const TransitionHydrationLane = /*                */ 0b0000000000000000000000001000000;
const TransitionLanes = /*                       */ 0b0000000011111111111111110000000;
const TransitionLane1 = /*                        */ 0b0000000000000000000000010000000;
const TransitionLane2 = /*                        */ 0b0000000000000000000000100000000;
const TransitionLane3 = /*                        */ 0b0000000000000000000001000000000;
const TransitionLane4 = /*                        */ 0b0000000000000000000010000000000;
const TransitionLane5 = /*                        */ 0b0000000000000000000100000000000;
const TransitionLane6 = /*                        */ 0b0000000000000000001000000000000;
const TransitionLane7 = /*                        */ 0b0000000000000000010000000000000;
const TransitionLane8 = /*                        */ 0b0000000000000000100000000000000;
const TransitionLane9 = /*                        */ 0b0000000000000001000000000000000;
const TransitionLane10 = /*                       */ 0b0000000000000010000000000000000;
const TransitionLane11 = /*                       */ 0b0000000000000100000000000000000;
const TransitionLane12 = /*                       */ 0b0000000000001000000000000000000;
const TransitionLane13 = /*                       */ 0b0000000000010000000000000000000;
const TransitionLane14 = /*                       */ 0b0000000000100000000000000000000;
const TransitionLane15 = /*                       */ 0b0000000001000000000000000000000;
const TransitionLane16 = /*                       */ 0b0000000010000000000000000000000;
const RetryLanes = /*                            */ 0b0000111100000000000000000000000;
const RetryLane1 = /*                             */ 0b0000000100000000000000000000000;
const RetryLane2 = /*                             */ 0b0000001000000000000000000000000;
const RetryLane3 = /*                             */ 0b0000010000000000000000000000000;
const RetryLane4 = /*                             */ 0b0000100000000000000000000000000;
exports.SomeRetryLane = RetryLane1;
exports.SelectiveHydrationLane = 0b0001000000000000000000000000000;
const NonIdleLanes = /*                          */ 0b0001111111111111111111111111111;
exports.IdleHydrationLane = 0b0010000000000000000000000000000;
exports.IdleLane = 0b0100000000000000000000000000000;
exports.OffscreenLane = 0b1000000000000000000000000000000;
exports.DiscreteEventPriority = exports.SyncLane; // 离散事件优先级(切片事件) 最高级
exports.ContinuousEventPriority = exports.InputContinuousLane; // 持续事件   次高级
exports.DefaultEventPriority = exports.DefaultLane; // 默认事件优先级
exports.IdleEventPriority = exports.IdleLane; // 闲置事件优先级
let currentUpdatePriority = exports.NoLane;
console.log('DiscreteEventPriority', exports.DiscreteEventPriority);
console.log('ContinuousEventPriority', exports.ContinuousEventPriority);
console.log('DefaultEventPriority', exports.DefaultEventPriority);
console.log('IdleEventPriority', exports.IdleEventPriority);
