// lzy-React/out/V3/myEvent/priority.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setCurrentUpdatePriority = exports.getCurrentUpdatePriority = exports.IdleEventPriority = exports.DefaultEventPriority = exports.ContinuousEventPriority = exports.DiscreteEventPriority = void 0;
const FiberLane_1 = require("./FiberLane");
//todo 维护了一个全局当前的优先级,会根据任务执行实时变化
let currentUpdatePriority = FiberLane_1.NoLane;
exports.DiscreteEventPriority = FiberLane_1.SyncLane; // 离散事件优先级(切片事件) 最高级
exports.ContinuousEventPriority = FiberLane_1.InputContinuousLane; // 持续事件   次高级
exports.DefaultEventPriority = FiberLane_1.DefaultLane; // 默认事件优先级
exports.IdleEventPriority = FiberLane_1.IdleLane; // 闲置事件优先级
function getCurrentUpdatePriority() {
    return currentUpdatePriority;
}
exports.getCurrentUpdatePriority = getCurrentUpdatePriority;
function setCurrentUpdatePriority(newPriority) {
    currentUpdatePriority = newPriority;
}
exports.setCurrentUpdatePriority = setCurrentUpdatePriority;
console.log('DiscreteEventPriority', exports.DiscreteEventPriority);
console.log('ContinuousEventPriority', exports.ContinuousEventPriority);
console.log('DefaultEventPriority', exports.DefaultEventPriority);
console.log('IdleEventPriority', exports.IdleEventPriority);
