// lzy-React/out/V3/myHook/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.myUseImperativeHandle = exports.createContext = exports.myUseContext = exports.myUseRef = exports.myUseEffect = exports.myUseState = void 0;
const useState_1 = require("./useState");
Object.defineProperty(exports, "myUseState", { enumerable: true, get: function () { return useState_1.myUseState; } });
const useEffect_1 = require("./useEffect");
Object.defineProperty(exports, "myUseEffect", { enumerable: true, get: function () { return useEffect_1.myUseEffect; } });
const useRef_1 = require("./useRef");
Object.defineProperty(exports, "myUseRef", { enumerable: true, get: function () { return useRef_1.myUseRef; } });
const useContext_1 = require("./useContext");
Object.defineProperty(exports, "myUseContext", { enumerable: true, get: function () { return useContext_1.myUseContext; } });
Object.defineProperty(exports, "createContext", { enumerable: true, get: function () { return useContext_1.createContext; } });
const useImperativeHandle_1 = require("./useImperativeHandle");
Object.defineProperty(exports, "myUseImperativeHandle", { enumerable: true, get: function () { return useImperativeHandle_1.myUseImperativeHandle; } });
