// lzy-React/lib/V3/myEvent/eventRegister.ts
import { FiberNode } from '../myReactCore/GlobalFiber'
import type { DOMEventName } from './domEventsName'
import { DOMEventNameList } from './domEventsName'
import {
    EventPriority,
    DiscreteEventPriority,
    ContinuousEventPriority,
    DefaultEventPriority,
    IdleEventPriority,
    getCurrentUpdatePriority,
    setCurrentUpdatePriority
} from './EventPriority'

// 需要注册的基础事件名



type ContainerElement = Document | Element | DocumentFragment
type EventListener = (e: Event) => any
export type AnyNativeEvent = Event | KeyboardEvent | MouseEvent | TouchEvent;

export const topLevelEventsToReactNames: Map<DOMEventName, string | null> =
    new Map();


// 注册React事件名  生成Map{'click'：'onClick'}
function registerSimpleEvents() {
    for (let i = 0; i < DOMEventNameList.length; i++) {
        const domEventName: DOMEventName = DOMEventNameList[i]; // click
        const capitalizedEvent = domEventName[0].toUpperCase() + domEventName.slice(1);//Click(大写)
        const ReactEventName = "on" + capitalizedEvent

        topLevelEventsToReactNames.set(domEventName, ReactEventName);
    }

}
registerSimpleEvents()




// 启动监听所有原生事件
export function listenToAllSupportedEvents(rootContainerElement: ContainerElement) {
    DOMEventNameList.forEach((domEventName) => {
        listenToNativeEvent(domEventName, true, rootContainerElement)
    })
    console.log(topLevelEventsToReactNames);
}

function listenToNativeEvent(domEventName: DOMEventName, isCapture: boolean, rootContainerElement: ContainerElement) {

    // 给事件进行优先级排序,创建真正的dispatch执行对象
    const eventPriority = getEventPriority(domEventName);
    console.log(eventPriority, domEventName, '---');

    // 根据优先级,创建listener, 进行绑定
    let listenerWrapper;
    switch (eventPriority) {
        case DiscreteEventPriority:
            listenerWrapper = dispatchDiscreteEvent; //最高级
            break;
        case ContinuousEventPriority:
            listenerWrapper = dispatchContinuousEvent;// 次高级
            break;
        case DefaultEventPriority:                   // 普通级
        default:
            listenerWrapper = dispatchEvent;         // 等待级
            break;
    }

    let listener = listenerWrapper.bind(null, domEventName, null, rootContainerElement)

    if (isCapture) {
        addEventCaptureListener(rootContainerElement, domEventName, listener);
    } else {
        addEventBubbleListener(rootContainerElement, domEventName, listener);
    }
}

export function removeEventListener(
    target: EventTarget,
    eventType: string,
    listener: EventListener,
    capture: boolean,
): void {
    target.removeEventListener(eventType, listener, capture);
}

export function addEventBubbleListener(
    target: EventTarget,
    eventType: string,
    listener: EventListener,
): Function {
    target.addEventListener(eventType, listener, false);
    return listener;
}

export function addEventCaptureListener(
    target: EventTarget,
    eventType: string,
    listener: EventListener,
): Function {
    target.addEventListener(eventType, listener, true);
    return listener;
}


// 将不同event进行分类
export function getEventPriority(domEventName: DOMEventName): EventPriority {
    switch (domEventName) {
        //! 无return时会返回2  也就是DiscreteEventPriority 
        //! 绝大部分用户事件优先级为2 最高级
        case 'cancel':
        case 'click':
        case 'close':
        case 'contextmenu':
        case 'copy':
        case 'cut':
        case 'auxclick':
        case 'dblclick':
        case 'dragend':
        case 'dragstart':
        case 'drop':
        case 'focusin':
        case 'focusout':
        case 'input':
        case 'invalid':
        case 'keydown':
        case 'keypress':
        case 'keyup':
        case 'mousedown':
        case 'mouseup':
        case 'paste':
        case 'pause':
        case 'play':
        case 'pointercancel':
        case 'pointerdown':
        case 'pointerup':
        case 'ratechange':
        case 'reset':
        case 'resize':
        case 'seeked':
        case 'submit':
        case 'touchcancel':
        case 'touchend':
        case 'touchstart':
        case 'volumechange':
        // Used by polyfills:
        // eslint-disable-next-line no-fallthrough
        case 'change':
        case 'selectionchange':
        case 'textInput':
        case 'compositionstart':
        case 'compositionend':
        case 'compositionupdate':
        // Only enableCreateEventHandleAPI:
        // eslint-disable-next-line no-fallthrough
        case 'beforeblur':
        case 'afterblur':
        // Not used by React but could be by user code:
        // eslint-disable-next-line no-fallthrough
        case 'beforeinput':
        case 'blur':
        case 'fullscreenchange':
        case 'focus':
        case 'hashchange':
        case 'popstate':
        case 'select':
        case 'selectstart':
            return DiscreteEventPriority;
        case 'drag':
        case 'dragenter':
        case 'dragexit':
        case 'dragleave':
        case 'dragover':
        case 'mousemove':
        case 'mouseout':
        case 'mouseover':
        case 'pointermove':
        case 'pointerout':
        case 'pointerover':
        case 'scroll':
        case 'toggle':
        case 'touchmove':
        case 'wheel':
        // Not used by React but could be by user code:
        // eslint-disable-next-line no-fallthrough
        case 'mouseenter':
        case 'mouseleave':
        case 'pointerenter':
        case 'pointerleave':
            return ContinuousEventPriority;
        case 'message':
            return DiscreteEventPriority;
        default:
            return DefaultEventPriority
    }
}

// 触发普通事件
function dispatchEvent(
    domEventName: DOMEventName,
    nativeEvent: AnyNativeEvent,
    targetInst: null | FiberNode,
    targetContainer: ContainerElement
) {
    console.log('触发dispatchEvent');

    const dispatchQueue = []
    //1. 提取所有监听的处理函数放到 dispatchQueue 当中
    // extractEvents()

    //2.模拟捕获阶段和冒泡阶段的执行流程，去执行所有的监听处理函数。
    // processDispatchQueue(dispatchQueue, eventSystemFlags);
}

//todo 触发DiscreteEvent离散事件(最高优先级)  用户事件级别
// 全局维护了一个当前时间的优先级CurrentUpdatePriority, 默认为default级
// 触发click事件时, 修改当前优先级为Discrete,事件执行完毕后还原
function dispatchDiscreteEvent(
    domEventName: DOMEventName,
    targetContainer: ContainerElement,
    nativeEvent: AnyNativeEvent,
) {
    const previousPriority = getCurrentUpdatePriority();

    try {
        setCurrentUpdatePriority(DiscreteEventPriority);
        dispatchEvent(domEventName, nativeEvent, null, targetContainer);
    } finally {
        setCurrentUpdatePriority(previousPriority); // 还原当前优先级
    }
}

//todo 触发ContinuousEvent持续事件(次高优先级)
function dispatchContinuousEvent(
    domEventName: DOMEventName,
    targetContainer: ContainerElement,
    nativeEvent: AnyNativeEvent,
) {
    const previousPriority = getCurrentUpdatePriority();

    try {
        setCurrentUpdatePriority(ContinuousEventPriority);
        dispatchEvent(domEventName, nativeEvent, null, targetContainer);
    } finally {
        setCurrentUpdatePriority(previousPriority); // 还原当前优先级
    }
}










// 首次render/createRoot时执行
// listenToAllSupportedEvents(rootContainerElement);
