// lzy-React/lib/V3/myEvent/eventRegister.ts

import type { DOMEventName } from './domEventsName'
import { DOMEventNameList } from './domEventsName'

import {
    EventPriority,
    DiscreteEventPriority,
    ContinuousEventPriority,
    DefaultEventPriority,
} from './EventPriority'


type EventListener = (e: Event) => any
export type ContainerElement = Document | Element | DocumentFragment
export type AnyNativeEvent = Event | KeyboardEvent | MouseEvent | TouchEvent;
export const topLevelEventsToReactNames: Map<DOMEventName, string | null> =
    new Map();

// 注册React事件名  生成Map{'click'：'onClick'}
function registerSimpleEvents() {
    for (let i = 0; i < DOMEventNameList.length; i++) {
        const domEventName: DOMEventName = DOMEventNameList[i]; // click
        const capitalizedEvent = domEventName[0].toUpperCase() + domEventName.slice(1);//Click(大写)
        const ReactEventName = "on" + capitalizedEvent

        topLevelEventsToReactNames.set(domEventName, ReactEventName);
    }

}

// 启动监听所有原生事件
export function listenToAllSupportedEvents(rootContainerElement: ContainerElement) {
    DOMEventNameList.forEach((domEventName) => {
        listenToNativeEvent(domEventName, true, rootContainerElement)
    })
    console.log(topLevelEventsToReactNames);
}

function listenToNativeEvent(domEventName: DOMEventName, isCapture: boolean, rootContainerElement: ContainerElement) {

    // 给事件进行优先级排序,创建真正的dispatch执行对象
    const eventPriority = getEventPriority(domEventName);

    // 根据优先级,创建listener, 进行绑定
    let listenerWrapper;
    switch (eventPriority) {
        case DiscreteEventPriority:
            listenerWrapper = dispatchDiscreteEvent; //最高级
            break;
        case ContinuousEventPriority:
            listenerWrapper = dispatchContinuousEvent;// 次高级
            break;
        case DefaultEventPriority:                   // 普通级
        default:
            listenerWrapper = dispatchEvent;         // 等待级
            break;
    }

    // 在执行listener时,event会作为第四个参数传入,也就是nativeEvnet
    let listener = listenerWrapper.bind(null, domEventName, rootContainerElement)

    if (isCapture) {
        addEventCaptureListener(rootContainerElement, domEventName, listener);
    } else {
        addEventBubbleListener(rootContainerElement, domEventName, listener);
    }
}

export function removeEventListener(
    target: EventTarget,
    eventType: string,
    listener: EventListener,
    capture: boolean,
): void {
    target.removeEventListener(eventType, listener, capture);
}

export function addEventBubbleListener(
    target: EventTarget,
    eventType: string,
    listener: EventListener,
): Function {
    target.addEventListener(eventType, listener, false);
    return listener;
}

export function addEventCaptureListener(
    target: EventTarget,
    eventType: string,
    listener: EventListener,
): Function {
    target.addEventListener(eventType, listener, true);
    return listener;
}


// 将不同event进行分类
export function getEventPriority(domEventName: DOMEventName): EventPriority {
    switch (domEventName) {
        //! 无return时会返回2  也就是DiscreteEventPriority 
        //! 绝大部分用户事件优先级为2 最高级
        case 'cancel':
        case 'click':
        case 'close':
        case 'contextmenu':
        case 'copy':
        case 'cut':
        case 'auxclick':
        case 'dblclick':
        case 'dragend':
        case 'dragstart':
        case 'drop':
        case 'focusin':
        case 'focusout':
        case 'input':
        case 'invalid':
        case 'keydown':
        case 'keypress':
        case 'keyup':
        case 'mousedown':
        case 'mouseup':
        case 'paste':
        case 'pause':
        case 'play':
        case 'pointercancel':
        case 'pointerdown':
        case 'pointerup':
        case 'ratechange':
        case 'reset':
        case 'resize':
        case 'seeked':
        case 'submit':
        case 'touchcancel':
        case 'touchend':
        case 'touchstart':
        case 'volumechange':
        // Used by polyfills:
        // eslint-disable-next-line no-fallthrough
        case 'change':
        case 'selectionchange':
        case 'textInput':
        case 'compositionstart':
        case 'compositionend':
        case 'compositionupdate':
        // Only enableCreateEventHandleAPI:
        // eslint-disable-next-line no-fallthrough
        case 'beforeblur':
        case 'afterblur':
        // Not used by React but could be by user code:
        // eslint-disable-next-line no-fallthrough
        case 'beforeinput':
        case 'blur':
        case 'fullscreenchange':
        case 'focus':
        case 'hashchange':
        case 'popstate':
        case 'select':
        case 'selectstart':
            return DiscreteEventPriority;
        case 'drag':
        case 'dragenter':
        case 'dragexit':
        case 'dragleave':
        case 'dragover':
        case 'mousemove':
        case 'mouseout':
        case 'mouseover':
        case 'pointermove':
        case 'pointerout':
        case 'pointerover':
        case 'scroll':
        case 'toggle':
        case 'touchmove':
        case 'wheel':
        // Not used by React but could be by user code:
        // eslint-disable-next-line no-fallthrough
        case 'mouseenter':
        case 'mouseleave':
        case 'pointerenter':
        case 'pointerleave':
            return ContinuousEventPriority;
        case 'message':
            return DiscreteEventPriority;
        default:
            return DefaultEventPriority
    }
}






// 首先注册所有的原生事件
registerSimpleEvents()

// 首次render/createRoot时启动监听所有原生事件
// listenToAllSupportedEvents(rootContainerElement);
