// lzy-React/lib/V3/myEvent/EventPriority.ts
import type { Lane } from './FiberLane'
import {
    NoLane,
    SyncLane,
    InputContinuousLane,
    DefaultLane,
    IdleLane,
} from './FiberLane';



//todo 维护了一个全局当前的优先级,会根据任务执行实时变化
export type EventPriority = Lane; // 事件优先级


let currentUpdatePriority: EventPriority = NoLane;

export const DiscreteEventPriority: EventPriority = SyncLane;               // 离散事件优先级(切片事件) 最高级
export const ContinuousEventPriority: EventPriority = InputContinuousLane;  // 持续事件   次高级
export const DefaultEventPriority: EventPriority = DefaultLane;             // 默认事件优先级
export const IdleEventPriority: EventPriority = IdleLane;                   // 闲置事件优先级

export function getCurrentUpdatePriority(): EventPriority {
    return currentUpdatePriority;
}

export function setCurrentUpdatePriority(newPriority: EventPriority) {
    currentUpdatePriority = newPriority;
}

console.log('DiscreteEventPriority', DiscreteEventPriority);
console.log('ContinuousEventPriority', ContinuousEventPriority);
console.log('DefaultEventPriority', DefaultEventPriority);
console.log('IdleEventPriority', IdleEventPriority);
