// lzy-React/out/index_V2.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createElement = exports.global = exports.Rekv = exports.updateRender = exports.render = exports.myUseEffect = exports.myUseState = void 0;
const useEffect_1 = require("./V2/myHook/useEffect");
Object.defineProperty(exports, "myUseEffect", { enumerable: true, get: function () { return useEffect_1.myUseEffect; } });
const useState_1 = require("./V2/myHook/useState");
Object.defineProperty(exports, "myUseState", { enumerable: true, get: function () { return useState_1.myUseState; } });
const render_1 = require("./V2/myReactCore/render");
Object.defineProperty(exports, "render", { enumerable: true, get: function () { return render_1.render; } });
Object.defineProperty(exports, "updateRender", { enumerable: true, get: function () { return render_1.updateRender; } });
const GlobalFiber_1 = require("./V2/myReactCore/GlobalFiber");
Object.defineProperty(exports, "global", { enumerable: true, get: function () { return GlobalFiber_1.global; } });
const index_1 = __importDefault(require("./V2/myRekV/index"));
exports.Rekv = index_1.default;
const createElement_1 = require("./V2/myJSX/createElement");
Object.defineProperty(exports, "createElement", { enumerable: true, get: function () { return createElement_1.createElement; } });
