// lzy-React/out/index_V3_concurrent.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.global = exports.Rekv = exports.createElement = exports.updateRender_concurrent = exports.render_concurrent = exports.myUseEffect = exports.myUseState = void 0;
const useEffect_1 = require("./V2/myHook/useEffect");
Object.defineProperty(exports, "myUseEffect", { enumerable: true, get: function () { return useEffect_1.myUseEffect; } });
const useState_concurrent_1 = require("./V2/concurrent/useState_concurrent");
Object.defineProperty(exports, "myUseState", { enumerable: true, get: function () { return useState_concurrent_1.myUseState; } });
const render_concurrent_1 = require("./V2/concurrent/render_concurrent");
Object.defineProperty(exports, "render_concurrent", { enumerable: true, get: function () { return render_concurrent_1.render_concurrent; } });
Object.defineProperty(exports, "updateRender_concurrent", { enumerable: true, get: function () { return render_concurrent_1.updateRender_concurrent; } });
const GlobalFiber_1 = require("./V2/myReactCore/GlobalFiber");
Object.defineProperty(exports, "global", { enumerable: true, get: function () { return GlobalFiber_1.global; } });
const index_1 = __importDefault(require("./V2/myRekV/index"));
exports.Rekv = index_1.default;
const createElement_1 = require("./V2/myJSX/createElement");
Object.defineProperty(exports, "createElement", { enumerable: true, get: function () { return createElement_1.createElement; } });
