// lzy-React/out/index_V3.js
"use strict";
// 使用了时间切片的render
Object.defineProperty(exports, "__esModule", { value: true });
exports.renderRoute = exports.myUseEffect = exports.myUseState = exports.updateRender = exports.render = exports.createElement = void 0;
const createElement_1 = require("./V3/myJSX/createElement");
Object.defineProperty(exports, "createElement", { enumerable: true, get: function () { return createElement_1.createElement; } });
const render_1 = require("./V3/myReactCore/render");
Object.defineProperty(exports, "render", { enumerable: true, get: function () { return render_1.render; } });
Object.defineProperty(exports, "updateRender", { enumerable: true, get: function () { return render_1.updateRender; } });
Object.defineProperty(exports, "renderRoute", { enumerable: true, get: function () { return render_1.renderRoute; } });
const useState_1 = require("./V3/myHook/useState");
Object.defineProperty(exports, "myUseState", { enumerable: true, get: function () { return useState_1.myUseState; } });
const useEffect_1 = require("./V3/myHook/useEffect");
Object.defineProperty(exports, "myUseEffect", { enumerable: true, get: function () { return useEffect_1.myUseEffect; } });
