// lzy-React/out/index_V3.js
"use strict";
// 使用了时间切片的render
Object.defineProperty(exports, "__esModule", { value: true });
exports.createContext = exports.forwardRef = exports.myUseImperativeHandle = exports.myUseRef = exports.myUseContext = exports.myUseEffect = exports.myUseState = exports.renderRoute = exports.updateRender = exports.unmountAtRoot = exports.render = exports.createElement = void 0;
const createElement_1 = require("./V3/myJSX/createElement");
Object.defineProperty(exports, "createElement", { enumerable: true, get: function () { return createElement_1.createElement; } });
const render_1 = require("./V3/myReactCore/render");
Object.defineProperty(exports, "render", { enumerable: true, get: function () { return render_1.render; } });
Object.defineProperty(exports, "updateRender", { enumerable: true, get: function () { return render_1.updateRender; } });
Object.defineProperty(exports, "renderRoute", { enumerable: true, get: function () { return render_1.renderRoute; } });
Object.defineProperty(exports, "unmountAtRoot", { enumerable: true, get: function () { return render_1.unmountAtRoot; } });
const index_1 = require("./V3/apis/index");
Object.defineProperty(exports, "forwardRef", { enumerable: true, get: function () { return index_1.forwardRef; } });
Object.defineProperty(exports, "createContext", { enumerable: true, get: function () { return index_1.createContext; } });
const index_2 = require("./V3/myHook/index");
Object.defineProperty(exports, "myUseState", { enumerable: true, get: function () { return index_2.myUseState; } });
Object.defineProperty(exports, "myUseEffect", { enumerable: true, get: function () { return index_2.myUseEffect; } });
Object.defineProperty(exports, "myUseRef", { enumerable: true, get: function () { return index_2.myUseRef; } });
Object.defineProperty(exports, "myUseContext", { enumerable: true, get: function () { return index_2.myUseContext; } });
Object.defineProperty(exports, "myUseImperativeHandle", { enumerable: true, get: function () { return index_2.myUseImperativeHandle; } });
const LzyReact = {
    createElement: createElement_1.createElement, render: render_1.render, unmountAtRoot: render_1.unmountAtRoot, updateRender: render_1.updateRender, renderRoute: render_1.renderRoute,
    myUseState: index_2.myUseState, myUseEffect: index_2.myUseEffect, myUseContext: index_2.myUseContext, myUseRef: index_2.myUseRef, myUseImperativeHandle: index_2.myUseImperativeHandle,
    forwardRef: index_1.forwardRef, createContext: index_1.createContext
};
exports.default = LzyReact;
